xarray/coding/times.py

```diff
@@ -667,12 +669,68 @@ def _division(deltas, delta, floor):
     return num
 
 
+def _cast_to_dtype_if_safe(num: np.ndarray, dtype: np.dtype) -> np.ndarray:
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", message="overflow")
+        cast_num = np.asarray(num, dtype=dtype)
+
+    if np.issubdtype(dtype, np.integer):
+        if not (num == cast_num).all():
+            if np.issubdtype(num.dtype, np.floating):
+                raise ValueError(
+                    f"Not possible to cast all encoded times from "
+                    f"{num.dtype!r} to {dtype!r} without losing precision. "
+                    f"Consider modifying the units such that integer values "
+                    f"can be used, or removing the units and dtype encoding, "
+                    f"at which point xarray will make an appropriate choice."
+                )
+            else:
+                raise OverflowError(
+                    f"Not possible to cast encoded times from "
+                    f"{num.dtype!r} to {dtype!r} without overflow. Consider "
+                    f"removing the dtype encoding, at which point xarray will "
+                    f"make an appropriate choice, or explicitly switching to "
+                    "a larger integer dtype."
+                )
+    else:
+        if np.isinf(cast_num).any():
+            raise OverflowError(
+                f"Not possible to cast encoded times from {num.dtype!r} to "
+                f"{dtype!r} without overflow.  Consider removing the dtype "
+                f"encoding, at which point xarray will make an appropriate "
+                f"choice, or explicitly switching to a larger floating point "
+                f"dtype."
+            )
+
+    return cast_num
+
+
+@overload
 def encode_cf_datetime(
-    dates,
+    dates: np.ndarray,
     units: str | None = None,
     calendar: str | None = None,
     dtype: np.dtype | None = None,
 ) -> tuple[np.ndarray, str, str]:
+    ...
+
+
+@overload
+def encode_cf_datetime(
+    dates: T_ChunkedArray,
```

Comment:
These should all be `xarray.core.types.T_DuckArray`, unless you're specifically going to call something on them that requires `.chunks`.